android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt

```diff
@@ -114,44 +119,46 @@ constructor(
    * [RuleConfig] against the [Facts] populated by the provided FHIR [Resource] s available in the
    * [relatedResourcesMap] and the [baseResource].
    */
-  fun fireRule(
-    ruleConfigs: List<RuleConfig>,
+  @Suppress("UNCHECKED_CAST")
+  suspend fun fireRules(
+    rules: Rules,
     baseResource: Resource? = null,
     relatedResourcesMap: Map<String, List<Resource>> = emptyMap(),
   ): Map<String, Any> {
-    // Reset previously computed values and init facts
-    computedValuesMap.clear()
-    facts.apply {
-      clear()
-      put(FHIR_PATH, fhirPathDataExtractor)
-      put(DATA, computedValuesMap)
-      put(SERVICE, rulesEngineService)
-    }
-
-    val customRules = mutableSetOf<Rule>()
-    ruleConfigs.forEach { ruleConfig ->
-
-      // Create JEXL rule
-      val customRule: JexlRule =
-        JexlRule(jexlEngine)
-          .name(ruleConfig.name)
-          .description(ruleConfig.description)
-          .priority(ruleConfig.priority)
-          .`when`(ruleConfig.condition.ifEmpty { TRUE })
-
-      ruleConfig.actions.forEach { customRule.then(it) }
-      customRules.add(customRule)
-    }
-
-    // baseResource is a FHIR resource whereas relatedResources is a list of FHIR resources
-    if (baseResource != null) {
-      facts.put(baseResource.resourceType.name, baseResource)
+    return withContext(dispatcherProvider.io()) {
+      // Initialize new facts and fire rules in background
+      facts =
+        Facts().apply {
+          put(FHIR_PATH, fhirPathDataExtractor)
+          put(DATA, mutableMapOf<String, Any>())
+          put(SERVICE, rulesEngineService)
+          if (baseResource != null) {
+            put(baseResource.resourceType.name, baseResource)
+          }
+          relatedResourcesMap.forEach { put(it.key, it.value) }
+        }
+      val timeToFireRules = measureTimeMillis { rulesEngine.fire(rules, facts) }
+      Timber.d("Rule executed in $timeToFireRules millisecond(s)")
```

Comment:
do we want to wrap this in a feature flag
```suggestion
      if (TRACK_TIMING) {
        val timeToFireRules = measureTimeMillis { rulesEngine.fire(rules, facts) }
        Timber.d("Rule executed in $timeToFireRules millisecond(s)")
      } else {
        rulesEngine.fire(rules, facts)
      }
```
not sure, but I assume on compilation it will remove the conditional if TRACK_TIMING is a constant

Reply:
We can check if `BuildConfig.Debug` is true instead, this log only makes sense while debugging.